smart-py/objkt_swap.py

```diff
@@ -17,135 +38,336 @@ def __init__(self, objkt, hdao, manager, metadata, curate):
             genesis = sp.timestamp(0),
             curate = curate,
             locked = False
-            )
-    
+        )
+
     @sp.entry_point
     def genesis(self):
-        sp.verify((sp.sender == self.data.manager) & ~(self.data.locked))
+        sp.verify(
+            (
+                sp.sender == self.data.manager
+            ) & ~(
+                self.data.locked
+            )
+        )
+
+        # what's up with adding 45 days?
         self.data.genesis = (sp.now).add_days(45)
         self.data.locked = True
-    
+
     @sp.entry_point
     def update_manager(self, params):
-        sp.verify(sp.sender == self.data.manager)
+        sp.verify(
+            sp.sender == self.data.manager
+        )
+
         self.data.manager = params
-        
+
     @sp.entry_point
     def swap(self, params):
-        sp.verify((params.objkt_amount > 0))
-        self.fa2_transfer(self.data.objkt, sp.sender, sp.to_address(sp.self), params.objkt_id, params.objkt_amount)
-        self.data.swaps[self.data.swap_id] = sp.record(issuer=sp.sender, objkt_id=params.objkt_id, objkt_amount=params.objkt_amount, xtz_per_objkt=params.xtz_per_objkt)
+        # must be at least one edition
+        sp.verify(params.objkt_amount > 0)
+        # no more than 10k
+        sp.verify(params.objkt_amount <= 10000)
+        # the objkt id must be at least 1
+        # TODO make this at least the starting number (152 atm)
+        sp.verify(params.objkt_id > 0)
+        # the objkt number should not be beyond the max number
+        sp.verify(params.objkt_id <= self.data.objkt_id)
+
+        # get the objkt
+        # TODO how to get the details of the objkt???
+        # this is where i want to get the details of the objkt
+        # by id from its contract but i don't know how to do this
+        #
+        # my last attempt was something like this:
+        #
+        # c = sp.record(
+        #     token_id = params.objkt_id
+        # )
+        #
+        # sp.result(c.token_id)
+        # sp.verify(c.token_id == params.objkt_id)
+        # print(c.amount)
+        # sp.verify(c.amount > 0)
+
+        # sp.verify(params.objkt_id == c.token_id)
+        # sp.verify(params.objkt_amount <= c.amount)
+
+        sp.verify(params.objkt_id <= self.data.objkt_id)
+
+        # the swap must carry a value of at least 0
+        sp.verify(params.xtz_per_objkt >= sp.utils.nat_to_mutez(0))
+
+        # TODO verify sender has the objkt and the number of editions they
+        # have is no more than their max number
+
+        self.fa2_transfer(
```

Comment:
Not sure, but should this be moved to the end of the method in case of a reentrant attack?

Reply:
can you please provide further info?

Reply:
From: https://consensys.github.io/smart-contract-best-practices/known_attacks/
The reentrant exploit is when an external call is made with subsequent contract data updates, but not sure if it applies here.

Reply:
Just learned that Tezos smart contracts were designed to avoid all the reentrant bugs that can happen in Ethereum, so I think we are good.